File: PythonLeetcode/leetcodeM/856_ScoreOfParentheses.py
"""
# 856. Score of Parentheses


# Given a balanced parentheses string S, compute the score of the string based on the following rule:

# () has score 1
# AB has score A + B, where A and B are balanced parentheses strings.
# (A) has score 2 * A, where A is a balanced parentheses string.


# Example 1:

# Input: "()"
# Output: 1
# Example 2:

# Input: "(())"
# Output: 2
# Example 3:

# Input: "()()"
# Output: 2
# Example 4:

# Input: "(()(()))"
# Output: 6
"""


class ScoreOfParentheses:

    """
    Approach 1: Divide and Conquer
    Intuition

    Split the string into S = A + B where A and B are balanced parentheses strings, and A is the smallest possible non-empty prefix of S.

    Algorithm

    Call a balanced string primitive if it cannot be partitioned into two non-empty balanced strings.

    By keeping track of balance (the number of ( parentheses minus the number of ) parentheses),
    we can partition S into primitive substrings S = P_1 + P_2 + ... + P_n. Then, score(S) = score(P_1) + score(P_2) + ... + score(P_n), by definition.

    For each primitive substring (S[i], S[i+1], ..., S[k]), if the string is length 2, then the score of this string is 1.
    Otherwise, it's twice the score of the substring (S[i+1], S[i+2], ..., S[k-1]).

    Complexity Analysis

    Time Complexity: O(N^2), where NN is the length of S. An example worst case is (((((((....))))))).

    Space Complexity: O(N), the size of the implied call stack.

    Complexity Analysis

    Time Complexity: O(N^2),  where NN is the length of S. An example worst case is (((((((....))))))).

    Space Complexity: O(N), the size of the implied call stack.
    """
    def doit_divide_conquer(self, S):

        def F(i, j):

            ans = base = 0

            for c in range(i, j):

                base += 1 if S[c] == '(' else -1

                if base == 0:
                    if c - i == 1:
                        ans += 1
                    else:
                        ans += 2 * F(i+1, c)

                    i = c + 1

            return ans

        return F(0, len(S))

    """
    Approach 2: Stack
    Intuition and Algorithm
    
    Every position in the string has a depth - some number of matching parentheses surrounding it. 
    For example, the dot in (()(.())) has depth 2, because of these parentheses: (__(.__))
    
    Our goal is to maintain the score at the current depth we are on. When we see an opening bracket, 
    we increase our depth, and our score at the new depth is 0. When we see a closing bracket, 
    we add twice the score of the previous deeper part - except when counting (), which has a score of 1.
    
    For example, when counting (()(())), our stack will look like this:
    
    [0, 0] after parsing (
    [0, 0, 0] after (
    [0, 1] after )
    [0, 1, 0] after (
    [0, 1, 0, 0] after (
    [0, 1, 1] after )
    [0, 3] after )
    [6] after )
    """
    def doit_stack(self, S):
        stack = [0]
        # The score of the current frame

        for x in S:
            if x == '(':
                stack.append(0)
            else:
                v = stack.pop()
                stack[-1] += max(2 * v, 1)

        return stack.pop()

    """
    Approach 3: Count Cores
    Intuition
    
    The final sum will be a sum of powers of 2, as every core (a substring (), with score 1) 
    will have it's score multiplied by 2 for each exterior set of parentheses that contains that core.
    
    Algorithm
    
    Keep track of the balance of the string, as defined in Approach #1. For every ) that immediately follows a (, the answer is 1 << balance, 
    as balance is the number of exterior set of parentheses that contains this core.
    """
    def doit_count(self, S):
        ans = bal = 0
        for i, x in enumerate(S):
            if x == '(':
                bal += 1
            else:
                bal -= 1
                if S[i-1] == '(':
                    ans += 1 << bal
        return ans


if __name__ == '__main__':

    res = ScoreOfParentheses().doit("()")  # 1

    res = ScoreOfParentheses().doit("(())")  # 2

    res = ScoreOfParentheses().doit("()()")  # 2

    res = ScoreOfParentheses().doit("(()()())")  # 6

    res = ScoreOfParentheses().doit("(((()(()())((())))))")

File: cppLeetcode/leetcodeM/856_ScoreOfParentheses.cpp
/*
856. Score of Parentheses

Given a balanced parentheses string S, compute the score of the string based on the following rule:

() has score 1
AB has score A + B, where A and B are balanced parentheses strings.
(A) has score 2 * A, where A is a balanced parentheses string.
 

Example 1:

Input: "()"
Output: 1
Example 2:

Input: "(())"
Output: 2
Example 3:

Input: "()()"
Output: 2
Example 4:

Input: "(()(()))"
Output: 6
 

Note:

S is a balanced parentheses string, containing only ( and ).
2 <= S.length <= 50


*/
#include <stack>
#include <string>

using std::string;
using std::stack;

class ScoreOFParentheses {

    /*
        856.Score-of-Parentheses
        此题的本质是一道Parentheses Parse。规则如下：同一层级的元素需要相加。更高一层级的元素（括号内）在脱括号的时候乘以2.不过有一个特殊的规则，如果更高一层级的元素是空，那么脱括号的时候赋值为1.
    */
    int doit_stack_parenthesis(string S) {
        stack<int>Stack;
        int cur = 0;
        for (auto ch: S)
        {
            if (ch==')')
            {
                if (cur==0)
                    cur = 1;
                else // "(((())))"
                    cur = cur*2;
                cur = Stack.top()+cur;
                Stack.pop();
            }
            else
            {
                // push it "((()))((..."
                Stack.push(cur);
                cur = 0;
            }
        }
        return cur;
    }

public:
    
    int doit_stack_parenthesis(string S) {
        
        stack<int> st;
        st.push(0);

        for (auto c : S) {

            if (c == '(')
                st.push(0);
            else {
                int v = st.top();
                st.pop();
                st.top() += std::max(2 * v, 1);
            }
        }
        
        return st.top();
    }

    /*
        Approach 0: Stack
        cur record the score at the current layer level.

        If we meet '(',
        we push the current score to stack,
        enter the next inner layer level,
        and reset cur = 0.

        If we meet ')',
        the cur score will be doubled and will be at least 1.
        We exit the current layer level,
        and set cur = stack.pop() + cur

        Complexity: O(N) time and O(N) space
    */
    int doit_stack_1(string S) {
        stack<int> stack;
        int cur = 0;
        for (char i : S)
            if (i == '(') {
                stack.push(cur);
                cur = 0;
            }
            else {
                cur += stack.top() + std::max(cur, 1);
                stack.pop();
            }
        return cur;
    }

    /*
        Approach 1: Array
        Same as stack, I do it with an array.
        We change a pointer instead of pushing/popping repeatedly.

        Complexity: O(N) time and O(N) space
    */
    int doit_array(string S) {
        int res[30] = {0}, i = 0;
        for (char c : S)
            if (c == '(') 
                res[++i] = 0;
            else 
                res[i - 1] += std::max(res[i] * 2, 1), i--;
        return res[0];
    }

    /*
        Follow-Up
        Can you solve it in O(1) space?

        Approach 2: O(1) Space
        We count the number of layers.
        If we meet '(' layers number l++
        else we meet ')' layers number l--

        If we meet "()", we know the number of layer outside,
        so we can calculate the score res += 1 << l.
    */
    int doit_01(string S) {
        
        int res = 0, l = 0;
        
        for (int i = 0; i < S.length(); ++i) {
            if (S[i] == '(') 
                l++; 
            else 
                l--;
            
            if (S[i] == ')' && S[i - 1] == '(') 
                res += 1 << l;
        }
        return res;
    }
};